src/librustc_trans/cabi_mips64.rs

```diff
@@ -8,50 +8,165 @@
+        // float fields.
+
+        if let layout::FieldPlacement::Arbitrary { .. } = ret.layout.fields {
+            let float_regs: Vec<Reg> =
+                (0..ret.layout.fields.count()).flat_map(|i| {
+                    match ret.layout.field(cx, i).abi {
+                        layout::Abi::Scalar(ref scalar) => match scalar.value {
+                            layout::F32 => Some(Reg::f32()),
+                            layout::F64 => Some(Reg::f64()),
+                            _ => None
+                        },
+                        _ => None
+                    }
+                }).collect();
+
+            if ret.layout.fields.count() == float_regs.len() {
+                if float_regs.len() == 1 {
+                    ret.cast_to(Uniform { unit: float_regs[0], total: float_regs[0].size });
+                    return;
+                } else if float_regs.len() == 2 {
+                    ret.cast_to(CastTarget::Pair(float_regs[0], float_regs[1]));
+                    return;
+                }
+            }
+        }
+
+        // Cast to a uniform int structure
+        ret.cast_to(Uniform {
+            unit: bits_to_int_reg(bits),
+            total: size
+        });
     } else {
         ret.make_indirect();
-        *offset += cx.tcx.data_layout.pointer_size;
     }
 }
 
-fn classify_arg_ty(cx: &CodegenCx, arg: &mut ArgType, offset: &mut Size) {
+fn classify_arg_ty<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>, arg: &mut ArgType<'tcx>) {
     let dl = &cx.tcx.data_layout;
     let size = arg.layout.size;
-    let align = arg.layout.align.max(dl.i32_align).min(dl.i64_align);
 
     if arg.layout.is_aggregate() {
-        arg.cast_to(Uniform {
-            unit: Reg::i64(),
-            total: size
-        });
-        if !offset.is_abi_aligned(align) {
-            arg.pad_with(Reg::i64());
+        let mut chunks: Vec<Reg> = Vec::new();
+        let trailing_int_size = match arg.layout.fields {
+            layout::FieldPlacement::Array { .. } => {
+                // Arrays are passed indirectly
+                arg.make_indirect();
+                return;
+            }
+            layout::FieldPlacement::Union(_) => {
+                // Unions and are always treated as a series of 64-bit integer chunks
+                size
+            },
+            layout::FieldPlacement::Arbitrary { .. } => {
+                // Structures are split up into a series of 64-bit integer chunks, but any aligned
+                // doubles not part of another aggregate are passed as floats.
+                let mut last_offset = Size::from_bytes(0);
+
+                for i in 0..arg.layout.fields.count() {
+                    let field = arg.layout.field(cx, i);
+                    let offset = arg.layout.fields.offset(i);
+
+                    // We only care about aligned doubles
+                    if let layout::Abi::Scalar(ref scalar) = field.abi {
+                        if let layout::F64 = scalar.value {
```

Comment:
How come this isn't recursive?

Reply:
As in why doesn't it recurse into structures within structures?

I originally thought this from reading the ABI document, but it's a bit ambiguous and it seems the ABI is intentionally not recursive. Eg this structure is passed in an integer register by clang and gcc even though it is an aligned double.
```c
struct test {
  struct {
    double x;
  };
};
```